File: pkg/templates/typescript/moondream-groq-computer-use/loop.ts
/**
 * Moondream computer-use agent loop.
 */

import type { Kernel } from '@onkernel/sdk';
import { ComputerTool } from './tools/computer';
import {
  ComputerAction,
  COORDINATE_SCALE,
  DEFAULT_SCREEN_SIZE,
  type ScreenSize,
} from './tools/types/computer';
import type { MoondreamClient, MoondreamPoint } from './moondream';

const URL_RE = /https?:\/\/[^\s)]+/i;

export interface StepInput {
  action: string;
  url?: string;
  target?: string;
  text?: string;
  question?: string;
  direction?: 'up' | 'down' | 'left' | 'right';
  magnitude?: number;
  x?: number;
  y?: number;
  keys?: string;
  seconds?: number;
  retries?: number;
  retry_delay_ms?: number;
  pre_wait_ms?: number;
  press_enter?: boolean;
  clear_before_typing?: boolean;
  length?: 'short' | 'normal' | 'long';
}

export interface AgentOptions {
  maxRetries?: number;
  retryDelayMs?: number;
  strict?: boolean;
}

export interface AgentResult {
  finalResponse: string;
  error?: string;
}

interface StepLog {
  step: number;
  action: string;
  status: 'success' | 'failed';
  detail: string;
  output?: string;
}

export async function runAgent({
  query,
  steps,
  moondream,
  kernel,
  sessionId,
  options,
}: {
  query?: string;
  steps?: StepInput[];
  moondream: MoondreamClient;
  kernel: Kernel;
  sessionId: string;
  options: AgentOptions;
}): Promise<AgentResult> {
  const computer = new ComputerTool(kernel, sessionId);
  const parsedSteps = steps?.length ? steps : parseSteps(query || '');

  if (!parsedSteps.length) {
    throw new Error('No steps could be derived from the query. Provide steps or a query.');
  }

  const logs: StepLog[] = [];
  const answers: string[] = [];
  let lastScreenshot: string | undefined;
  let error: string | undefined;

  for (const [index, step] of parsedSteps.entries()) {
    const stepNumber = index + 1;
    const action = (step.action || '').trim().toLowerCase();

    if (!action) {
      logs.push({ step: stepNumber, action: 'unknown', status: 'failed', detail: 'Missing action' });
      if (options.strict) {
        error = 'Missing action in step';
        break;
      }
      continue;
    }

    try {
      if (step.pre_wait_ms) {
        await sleep(step.pre_wait_ms);
      }

      if (action === 'open_web_browser' || action === 'open') {
        const result = await computer.executeAction(ComputerAction.OPEN_WEB_BROWSER, {});
        lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        logs.push({ step: stepNumber, action, status: status(result.error), detail: 'Opened browser' });
      } else if (action === 'navigate') {
        const url = step.url || findUrl(query || '');
        if (!url) throw new Error('navigate requires url');
        const result = await computer.executeAction(ComputerAction.NAVIGATE, { url });
        lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        logs.push({ step: stepNumber, action, status: status(result.error), detail: `Navigated to ${url}` });
      } else if (action === 'go_back') {
        const result = await computer.executeAction(ComputerAction.GO_BACK, {});
        lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        logs.push({ step: stepNumber, action, status: status(result.error), detail: 'Went back' });
      } else if (action === 'go_forward') {
        const result = await computer.executeAction(ComputerAction.GO_FORWARD, {});
        lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        logs.push({ step: stepNumber, action, status: status(result.error), detail: 'Went forward' });
      } else if (action === 'search') {
        const result = await computer.executeAction(ComputerAction.SEARCH, {});
        lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        logs.push({ step: stepNumber, action, status: status(result.error), detail: 'Focused address bar' });
      } else if (action === 'wait') {
        const seconds = step.seconds ?? 1;
        await sleep(seconds * 1000);
        logs.push({ step: stepNumber, action, status: 'success', detail: `Waited ${seconds.toFixed(2)}s` });
      } else if (action === 'key') {
        if (!step.keys) throw new Error('key action requires keys');
        const result = await computer.executeAction(ComputerAction.KEY_COMBINATION, { keys: step.keys });
        lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        logs.push({ step: stepNumber, action, status: status(result.error), detail: `Pressed ${step.keys}` });
      } else if (action === 'scroll') {
        const direction = step.direction ?? 'down';
        const magnitude = step.magnitude;
        if (step.x !== undefined && step.y !== undefined) {
          const [xNorm, yNorm] = normalizePoint(step.x, step.y, computer.getScreenSize());
          const args: Record<string, number | string> = { x: xNorm, y: yNorm, direction };
          if (magnitude !== undefined) args.magnitude = magnitude;
          const result = await computer.executeAction(ComputerAction.SCROLL_AT, args);
          lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        } else {
          const args: Record<string, number | string> = { direction };
          if (magnitude !== undefined) args.magnitude = magnitude;
          const result = await computer.executeAction(ComputerAction.SCROLL_DOCUMENT, args);
          lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
        }
        logs.push({ step: stepNumber, action, status: 'success', detail: `Scrolled ${direction}` });
      } else if (action === 'click' || action === 'type') {
        const target = step.target;
        const retries = step.retries ?? options.maxRetries ?? 3;
        const delayMs = step.retry_delay_ms ?? options.retryDelayMs ?? 1000;

        const coords = await resolveTargetCoords({
          step,
          target,
          moondream,
          computer,
          lastScreenshot,
          retries,
          delayMs,
        });

        if (!coords) throw new Error(`Unable to locate target: ${target}`);

        const [xNorm, yNorm] = coords;
        if (action === 'click') {
          const result = await computer.executeAction(ComputerAction.CLICK_AT, { x: xNorm, y: yNorm });
          lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
          logs.push({ step: stepNumber, action, status: status(result.error), detail: `Clicked ${target}` });
        } else {
          if (step.text === undefined) throw new Error('type action requires text');
          const result = await computer.executeAction(ComputerAction.TYPE_TEXT_AT, {
            x: xNorm,
            y: yNorm,
            text: String(step.text),
            press_enter: Boolean(step.press_enter),
            clear_before_typing: step.clear_before_typing !== false,
          });
          lastScreenshot = updateScreenshot(result.base64Image, lastScreenshot);
          logs.push({ step: stepNumber, action, status: status(result.error), detail: `Typed into ${target}` });
        }
      } else if (action === 'query') {
        const question = step.question || query;
        if (!question) throw new Error('query action requires question');
        const screenshot = await ensureScreenshot(computer, lastScreenshot);
        lastScreenshot = screenshot;
        const answer = await moondream.query(screenshot, String(question));
        answers.push(answer);
        logs.push({ step: stepNumber, action, status: 'success', detail: 'Answered question', output: answer });
      } else if (action === 'caption') {
        const length = step.length ?? 'normal';
        const screenshot = await ensureScreenshot(computer, lastScreenshot);
        lastScreenshot = screenshot;
        const caption = await moondream.caption(screenshot, length);
        answers.push(caption);
        logs.push({ step: stepNumber, action, status: 'success', detail: 'Generated caption', output: caption });
      } else {
        throw new Error(`Unknown action: ${action}`);
      }
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      logs.push({ step: stepNumber, action, status: 'failed', detail: message });
      error = message;
      if (options.strict) break;
    }
  }

  const summary = `Completed ${logs.filter(l => l.status === 'success').length}/${logs.length} steps`;
  const resultPayload = {
    summary,
    steps: logs,
    answers,
  };

  return {
    finalResponse: JSON.stringify(resultPayload, null, 2),
    error,
  };
}

export function parseSteps(query: string): StepInput[] {
  const trimmed = query.trim();
  if (!trimmed) return [];

  if (trimmed.startsWith('{') || trimmed.startsWith('[')) {
    try {
      const data = JSON.parse(trimmed);
      if (Array.isArray(data)) return data as StepInput[];
      if (data && typeof data === 'object' && Array.isArray((data as { steps?: StepInput[] }).steps)) {
        return (data as { steps: StepInput[] }).steps;
      }
    } catch {
      // fall through
    }
  }

  const steps: StepInput[] = [];
  const url = findUrl(trimmed);
  if (url) steps.push({ action: 'navigate', url });

  const question = stripUrlAndNavigation(trimmed);
  const wantsCaption = /\bdescribe|caption\b/i.test(trimmed);

  if (wantsCaption) {
    steps.push({ action: 'caption' });
  } else if (question) {
    steps.push({ action: 'query', question });
  } else if (url) {
    steps.push({ action: 'caption' });
  } else {
    steps.push({ action: 'query', question: trimmed });
  }

  return steps;
}

function findUrl(query: string): string | undefined {
  const match = query.match(URL_RE);
  return match?.[0];
}

function stripUrlAndNavigation(query: string): string {
  let cleaned = query.replace(URL_RE, '');
  cleaned = cleaned.replace(/\b(navigate|open|go|visit)\b/gi, '');
  cleaned = cleaned.replace(/\bto\b/gi, ' ');
  cleaned = cleaned.replace(/\s+/g, ' ').trim();
  return cleaned.replace(/^[,.;:-]+|[,.;:-]+$/g, '');
}

function normalizePoint(x: number, y: number, screenSize?: ScreenSize): [number, number] {
  if (x >= 0 && x <= 1 && y >= 0 && y <= 1) {
    return [Math.round(x * COORDINATE_SCALE), Math.round(y * COORDINATE_SCALE)];
  }
  const width = screenSize?.width ?? DEFAULT_SCREEN_SIZE.width;
  const height = screenSize?.height ?? DEFAULT_SCREEN_SIZE.height;
  return [
    Math.round((x / width) * COORDINATE_SCALE),
    Math.round((y / height) * COORDINATE_SCALE),
  ];
}

function updateScreenshot(current?: string, fallback?: string): string | undefined {
  return current ?? fallback;
}

function status(error?: string): 'success' | 'failed' {
  return error ? 'failed' : 'success';
}

async function ensureScreenshot(computer: ComputerTool, lastScreenshot?: string): Promise<string> {
  if (lastScreenshot) return lastScreenshot;
  const result = await computer.screenshot();
  if (result.error || !result.base64Image) {
    throw new Error(result.error || 'Failed to capture screenshot');
  }
  return result.base64Image;
}

async function resolveTargetCoords({
  step,
  target,
  moondream,
  computer,
  lastScreenshot,
  retries,
  delayMs,
}: {
  step: StepInput;
  target?: string;
  moondream: MoondreamClient;
  computer: ComputerTool;
  lastScreenshot?: string;
  retries: number;
  delayMs: number;
}): Promise<[number, number] | undefined> {
  if (step.x !== undefined && step.y !== undefined) {
    return normalizePoint(step.x, step.y, computer.getScreenSize());
  }

  if (!target) return undefined;

  const attempts = Math.max(1, retries);
  let currentScreenshot = lastScreenshot;

  for (let attempt = 0; attempt < attempts; attempt++) {
    const screenshot = await ensureScreenshot(computer, currentScreenshot);
    const point = await moondream.point(screenshot, String(target));
    if (point) {
      return normalizePoint(point.x, point.y, computer.getScreenSize());
    }
    if (attempt < attempts - 1) {
      await sleep(delayMs);
      currentScreenshot = undefined;
    }
  }

  return undefined;
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

File: pkg/templates/python/moondream-groq-computer-use/loop.py
"""
Moondream computer-use agent loop.
"""

from __future__ import annotations

import asyncio
import json
import re
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Optional, Tuple

from kernel import Kernel

from moondream import MoondreamClient
from tools import (
    ComputerAction,
    ComputerTool,
    COORDINATE_SCALE,
    DEFAULT_SCREEN_SIZE,
    ScreenSize,
)


URL_RE = re.compile(r"https?://[^\s)]+", re.IGNORECASE)


@dataclass
class AgentOptions:
    max_retries: int = 3
    retry_delay_ms: int = 1000
    strict: bool = False


@dataclass
class StepLog:
    step: int
    action: str
    status: str
    detail: str
    output: Optional[str] = None


async def run_agent(
    *,
    query: Optional[str],
    steps: Optional[List[Dict[str, Any]]],
    moondream: MoondreamClient,
    kernel: Kernel,
    session_id: str,
    options: AgentOptions,
) -> Dict[str, Any]:
    computer = ComputerTool(kernel, session_id)

    parsed_steps = steps or parse_steps(query or "")
    if not parsed_steps:
        raise ValueError("No steps could be derived from the query. Provide steps or a query.")

    logs: List[StepLog] = []
    answers: List[str] = []
    last_screenshot: Optional[str] = None
    error: Optional[str] = None

    for index, step in enumerate(parsed_steps, start=1):
        action = str(step.get("action", "")).strip().lower()
        if not action:
            logs.append(StepLog(index, "unknown", "failed", "Missing action"))
            if options.strict:
                error = "Missing action in step"
                break
            continue

        try:
            if step.get("pre_wait_ms"):
                await asyncio.sleep(float(step["pre_wait_ms"]) / 1000)

            if action in {"open_web_browser", "open"}:
                result = await computer.execute_action(ComputerAction.OPEN_WEB_BROWSER, {})
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), "Opened browser"))

            elif action == "navigate":
                url = step.get("url") or _find_url(query or "")
                if not url:
                    raise ValueError("navigate requires url")
                result = await computer.execute_action(ComputerAction.NAVIGATE, {"url": url})
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), f"Navigated to {url}"))

            elif action == "go_back":
                result = await computer.execute_action(ComputerAction.GO_BACK, {})
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), "Went back"))

            elif action == "go_forward":
                result = await computer.execute_action(ComputerAction.GO_FORWARD, {})
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), "Went forward"))

            elif action == "search":
                result = await computer.execute_action(ComputerAction.SEARCH, {})
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), "Focused address bar"))

            elif action == "wait":
                seconds = float(step.get("seconds", 1))
                await asyncio.sleep(seconds)
                logs.append(StepLog(index, action, "success", f"Waited {seconds:.2f}s"))

            elif action == "key":
                keys = step.get("keys")
                if not keys:
                    raise ValueError("key action requires keys")
                result = await computer.execute_action(ComputerAction.KEY_COMBINATION, {"keys": keys})
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), f"Pressed {keys}"))

            elif action == "scroll":
                direction = step.get("direction", "down")
                magnitude = step.get("magnitude")
                if "x" in step and "y" in step:
                    x_norm, y_norm = normalize_point(
                        float(step["x"]),
                        float(step["y"]),
                        computer.screen_size,
                    )
                    args: Dict[str, Any] = {
                        "x": x_norm,
                        "y": y_norm,
                        "direction": direction,
                    }
                    if magnitude is not None:
                        args["magnitude"] = int(magnitude)
                    result = await computer.execute_action(ComputerAction.SCROLL_AT, args)
                else:
                    args = {"direction": direction}
                    if magnitude is not None:
                        args["magnitude"] = int(magnitude)
                    result = await computer.execute_action(ComputerAction.SCROLL_DOCUMENT, args)
                last_screenshot = _update_screenshot(result, last_screenshot)
                logs.append(StepLog(index, action, _status(result), f"Scrolled {direction}"))

            elif action in {"click", "type"}:
                target = step.get("target")
                retries = int(step.get("retries", options.max_retries))
                delay_ms = int(step.get("retry_delay_ms", options.retry_delay_ms))

                coords = await _resolve_target_coords(
                    step,
                    target,
                    moondream,
                    computer,
                    last_screenshot,
                    retries,
                    delay_ms,
                )

                if not coords:
                    raise ValueError(f"Unable to locate target: {target}")

                x_norm, y_norm = coords
                if action == "click":
                    result = await computer.execute_action(
                        ComputerAction.CLICK_AT,
                        {"x": x_norm, "y": y_norm},
                    )
                    last_screenshot = _update_screenshot(result, last_screenshot)
                    logs.append(StepLog(index, action, _status(result), f"Clicked {target}"))
                else:
                    text = step.get("text")
                    if text is None:
                        raise ValueError("type action requires text")
                    result = await computer.execute_action(
                        ComputerAction.TYPE_TEXT_AT,
                        {
                            "x": x_norm,
                            "y": y_norm,
                            "text": str(text),
                            "press_enter": bool(step.get("press_enter", False)),
                            "clear_before_typing": bool(step.get("clear_before_typing", True)),
                        },
                    )
                    last_screenshot = _update_screenshot(result, last_screenshot)
                    logs.append(
                        StepLog(index, action, _status(result), f"Typed into {target}")
                    )

            elif action == "query":
                question = step.get("question") or query
                if not question:
                    raise ValueError("query action requires question")
                screenshot = await _ensure_screenshot(computer, last_screenshot)
                last_screenshot = screenshot
                answer = await moondream.query(screenshot, str(question))
                answers.append(answer)
                logs.append(StepLog(index, action, "success", "Answered question", output=answer))

            elif action == "caption":
                length = step.get("length", "normal")
                screenshot = await _ensure_screenshot(computer, last_screenshot)
                last_screenshot = screenshot
                caption = await moondream.caption(screenshot, str(length))
                answers.append(caption)
                logs.append(StepLog(index, action, "success", "Generated caption", output=caption))

            else:
                raise ValueError(f"Unknown action: {action}")

        except Exception as exc:
            message = str(exc)
            logs.append(StepLog(index, action, "failed", message))
            error = message
            if options.strict:
                break

    summary = f"Completed {sum(1 for log in logs if log.status == 'success')}/{len(logs)} steps"
    result_payload = {
        "summary": summary,
        "steps": [log.__dict__ for log in logs],
        "answers": answers,
    }

    return {
        "final_response": json.dumps(result_payload, indent=2),
        "error": error,
    }


def parse_steps(query: str) -> List[Dict[str, Any]]:
    query = query.strip()
    if not query:
        return []

    if query.startswith("{") or query.startswith("["):
        try:
            data = json.loads(query)
            if isinstance(data, list):
                return data
            if isinstance(data, dict) and isinstance(data.get("steps"), list):
                return data["steps"]
        except json.JSONDecodeError:
            pass

    steps: List[Dict[str, Any]] = []
    url = _find_url(query)
    if url:
        steps.append({"action": "navigate", "url": url})

    question = _strip_url_and_navigation(query)
    wants_caption = any(term in query.lower() for term in ["describe", "caption"])

    if wants_caption:
        steps.append({"action": "caption"})
    elif question:
        steps.append({"action": "query", "question": question})
    elif url:
        steps.append({"action": "caption"})
    else:
        steps.append({"action": "query", "question": query})

    return steps


def _find_url(query: str) -> Optional[str]:
    match = URL_RE.search(query)
    return match.group(0) if match else None


def _strip_url_and_navigation(query: str) -> str:
    cleaned = URL_RE.sub("", query)
    cleaned = re.sub(r"\b(navigate|open|go|visit)\b", "", cleaned, flags=re.IGNORECASE)
    cleaned = cleaned.replace("to", " ")
    cleaned = re.sub(r"\s+", " ", cleaned).strip(" ,.;:-")
    return cleaned


def normalize_point(x: float, y: float, screen_size: Optional[ScreenSize] = None) -> Tuple[int, int]:
    if 0 <= x <= 1 and 0 <= y <= 1:
        return int(x * COORDINATE_SCALE), int(y * COORDINATE_SCALE)
    width = screen_size.width if screen_size else DEFAULT_SCREEN_SIZE.width
    height = screen_size.height if screen_size else DEFAULT_SCREEN_SIZE.height
    return int((x / width) * COORDINATE_SCALE), int((y / height) * COORDINATE_SCALE)


def _status(result: Any) -> str:
    return "failed" if result and getattr(result, "error", None) else "success"


def _update_screenshot(result: Any, last_screenshot: Optional[str]) -> Optional[str]:
    if result and getattr(result, "base64_image", None):
        return result.base64_image
    return last_screenshot


async def _ensure_screenshot(computer: ComputerTool, last_screenshot: Optional[str]) -> str:
    if last_screenshot:
        return last_screenshot
    result = await computer.screenshot()
    if result.error or not result.base64_image:
        raise RuntimeError(result.error or "Failed to capture screenshot")
    return result.base64_image


async def _resolve_target_coords(
    step: Dict[str, Any],
    target: Optional[str],
    moondream: MoondreamClient,
    computer: ComputerTool,
    last_screenshot: Optional[str],
    retries: int,
    delay_ms: int,
) -> Optional[Tuple[int, int]]:
    if "x" in step and "y" in step:
        return normalize_point(float(step["x"]), float(step["y"]), computer.screen_size)

    if not target:
        return None

    attempts = max(1, retries)
    for attempt in range(attempts):
        screenshot = await _ensure_screenshot(computer, last_screenshot)
        point = await moondream.point(screenshot, str(target))
        if point:
            return normalize_point(point.x, point.y, computer.screen_size)
        if attempt < attempts - 1:
            await asyncio.sleep(delay_ms / 1000)
            last_screenshot = None

    return None
